src/main/java/com/gotocompany/depot/maxcompute/MaxComputeSink.java

```diff
@@ -0,0 +1,65 @@
+import com.gotocompany.depot.SinkResponse;
+import com.gotocompany.depot.error.ErrorInfo;
+import com.gotocompany.depot.error.ErrorType;
+import com.gotocompany.depot.exception.SinkException;
+import com.gotocompany.depot.maxcompute.client.MaxComputeClient;
+import com.gotocompany.depot.maxcompute.converter.record.MessageRecordConverter;
+import com.gotocompany.depot.maxcompute.model.RecordWrapper;
+import com.gotocompany.depot.maxcompute.model.RecordWrappers;
+import com.gotocompany.depot.message.Message;
+import com.gotocompany.depot.metrics.Instrumentation;
+import com.gotocompany.depot.metrics.MaxComputeMetrics;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.util.List;
+
+@RequiredArgsConstructor
+@Slf4j
+public class MaxComputeSink implements Sink {
+
+    private final MaxComputeClient maxComputeClient;
+    private final MessageRecordConverter messageRecordConverter;
+    private final Instrumentation instrumentation;
+    private final MaxComputeMetrics maxComputeMetrics;
+
+    @Override
+    public SinkResponse pushToSink(List<Message> messages) throws SinkException {
+        SinkResponse sinkResponse = new SinkResponse();
+        RecordWrappers recordWrappers = messageRecordConverter.convert(messages);
+        recordWrappers.getInvalidRecords()
+                .forEach(invalidRecord -> sinkResponse.getErrors().put(invalidRecord.getIndex(), invalidRecord.getErrorInfo()));
+        try {
+            maxComputeClient.insert(recordWrappers.getValidRecords());
+        } catch (IOException | TunnelException e) {
+            log.error("Error while inserting records to MaxCompute: ", e);
+            mapInsertionError(recordWrappers.getValidRecords(), sinkResponse, new ErrorInfo(e, ErrorType.SINK_RETRYABLE_ERROR));
+            instrumentation.incrementCounter(maxComputeMetrics.getMaxComputeOperationTotalMetric(),
+                    String.format(MaxComputeMetrics.MAXCOMPUTE_ERROR_TAG, e.getClass().getSimpleName()));
+        } catch (Exception e) {
+            log.error("Error while inserting records to MaxCompute: ", e);
+            mapInsertionError(recordWrappers.getValidRecords(), sinkResponse, new ErrorInfo(e, ErrorType.DEFAULT_ERROR));
+            instrumentation.incrementCounter(maxComputeMetrics.getMaxComputeOperationTotalMetric(),
+                    String.format(MaxComputeMetrics.MAXCOMPUTE_ERROR_TAG, e.getClass().getSimpleName()));
+        }
+        return sinkResponse;
+    }
+
+    @Override
+    public void close() throws IOException {
+    }
+
+    private void mapInsertionError(List<RecordWrapper> recordWrappers,
+                                   SinkResponse sinkResponse,
+                                   ErrorInfo errorInfo) {
+        recordWrappers.forEach(recordWrapper -> {
+            recordWrapper.setErrorInfo(errorInfo);
+            sinkResponse.getErrors().put(recordWrapper.getIndex(), errorInfo);
```

Comment:
The error mapping reuses the same ErrorInfo instance across all records, losing context about which specific record failed. Should create individual error instances with specific context.

Reply:
MaxCompute's StreamRecordPack behaves atomically, either all the records are all successfully inserted or none.